### PR TITLE
External api integration

### PR DESCRIPTION
**Pull request by daniel-vega-86**
## Summary

[Change!] 
- Install httparty
- add books route
- create books controller
- create books service

## Screenshots

[Change!] N/A.

## Trello Card

[Change!] https://trello.com/c/Pi1mwZyw/34-2-integraci%C3%B3n-api-externa

### CODE REVIEW

File: app/services/books_service.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+class BooksService
+  include HTTParty
+  base_uri 'openlibrary.org'
```
Comment (blacksam07) [inline app/services/books_service.rb]: use credentials in this case, read this article https://blog.saeloun.com/2019/10/10/rails-6-adds-support-for-multi-environment-credentials.html

File: config/routes.rb [hunk 2]
```diff
@@ -1,4 +1,10 @@
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+  resources :books do
```
Comment (blacksam07) [inline config/routes.rb]: Use the resources crud method read this https://guides.rubyonrails.org/routing.html#crud-verbs-and-actions
like 
```ruby
resources :books, only: :show
```

File: app/services/books_service.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+  def find_book(isbn)
+    options = { query: { bibkeys: isbn, format: 'json', jscmd: 'data' } }
+    book = self.class.get('/api/books', options)
+
+    keys = %w[title subtitle number_of_pages authors]
+    JSON.parse(book.body)[isbn].keep_if { |k, _| keys.include?(k) }
```
Comment (blacksam07) [inline app/services/books_service.rb]: ❤️

**Review by blacksam07 — APPROVED** [review]

**Review by blacksam07 — CHANGES_REQUESTED** [review]
little changes

File: app/services/books_service.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+class BooksService
+  include HTTParty
+  base_uri 'openlibrary.org'
+
+  def find_book(isbn)
+    options = { query: { bibkeys: isbn, format: 'json', jscmd: 'data' } }
+    book = self.class.get('/api/books', options)
```
Comment (redwarewolf) [inline app/services/books_service.rb]: The route can be saved in a constant preferrably

File: app/services/books_service.rb (hunk 3 above)
Comment (redwarewolf) [inline app/services/books_service.rb]: Use Slice for this, but nice idea.

Comment (redwarewolf) [reply app/services/books_service.rb]: https://apidock.com/rails/Hash/slice

**Review by redwarewolf — CHANGES_REQUESTED** [review]

File: app/services/books_service.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+class BooksService
+  include HTTParty
+  base_uri Rails.application.credentials.base_uri
+
+  def find_book(id)
+    path = '/api/books'
```
Comment (blacksam07) [inline app/services/books_service.rb]: move at constant class outside the method, and include `/api` in base_uri
```ruby
BOOKS_PATH = '/books'.freeze # the freeze method locks editing
```

File: app/services/books_service.rb [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+    options = { query: { bibkeys: id, format: 'json', jscmd: 'data' } }
+    book = self.class.get(path, options)
+
+    keys = %i[title subtitle number_of_pages authors]
+    JSON.parse(book.body)[id].deep_symbolize_keys.slice(*keys)
```
Comment (blacksam07) [inline app/services/books_service.rb]: move this logic to another method like at `process_response` the function of find_book method is only find not process a data

File: config/routes.rb [hunk 7]
```diff
@@ -1,5 +1,7 @@
 Rails.application.routes.draw do
   mount_devise_token_auth_for 'User', at: 'auth'
   # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+  resources :books, only: :show
```
Comment (blacksam07) [inline config/routes.rb]: since you are finding the book by ISBN, you can change the parameter
```ruby
resources :books, only: :show, param: :isbn
```

**Review by blacksam07 — APPROVED** [review]

**Review by blacksam07 — APPROVED** [review]

**Review by redwarewolf — APPROVED** [review]

**Pull request merged**